AY2021S1-CS2103-T16-3/tp
# Improve undo/redo SoC, fix UG and DG

### What this does
This PR refactors code related to undo/redo functionality, renaming `VersionedResiReg` to `StatefulResiReg` (to better convey the idea that ResiReg reacts to the same input (that is the undo and redo commands) differently depending on the current state).

It also changes the implementation to storing states in two separate stacks instead of using a list of past states, to follow the Separation of Concerns (SoC) principle so that processes related to `undo` and `redo` can be viewed separately in their own right for easier understanding. Additionally, in the event that there is a change in implementation of commands that modify state (from saving the state to each having its own `undo()` method), there will be less additional work done. 

Tests for `StatefulResiReg` have also been modified to support this new implementation (other tests remain unaltered since they use the changed methods as black boxes). 

The PR also updates the UG and DG to reflect new commands (`undo`, `redo` and `history`) and update information regarding the undo/redo implementation.


Part of #31, #33
Closes #120

### How to test

1. Type in a sequence of commands that change state (`add`, `allocate`, etc.), along with a combination of `undo` and `redo`
2. Ensure that the sequence of commands entered output the correct results
(**Note**: If a new command that changes state is entered while there is a command that can be redone, this latter command will be cleared, as documented in the DG).

### Notes

N/A
